pkg/chains/signing.go

```diff
@@ -247,8 +248,22 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 		}
 	}
 
-	// Now mark the TaskRun as signed
-	return MarkSigned(ctx, tektonObj, o.Pipelineclientset, extraAnnotations)
+	// Now mark the TektonObject as signed
+	if err := MarkSigned(ctx, tektonObj, o.Pipelineclientset, extraAnnotations); err != nil {
+		return err
+	}
+
+	// Notify parent object that child object has been signed. This allows PipelineRun resources
+	// to be reconciled when the status of its TaskRuns use ChildReferences.
+	if parent, err := tektonObj.GetParent(ctx, o.Pipelineclientset); err != nil {
+		return err
+	} else if parent != nil {
+		if err := MarkParentSigned(ctx, parent, o.Pipelineclientset); err != nil {
```

Comment:
Do we need to wait until all Tasks are done before doing this? 

In particular I'm thinking about if we have a matrix pipeline that builds images for different arch in different Tasks - if we set this too early the Pipeline sig might still clash with other Tasks. 

Maybe we invert this relationship and have objects query for their parent status? 🤔

Reply:
> Do we need to wait until all Tasks are done before doing this?

In reality, we just need to wait until the last Task within the Pipeline that has signable subjects. We could enhance the logic to handle that, not sure how much more beneficial that would be though.

> In particular I'm thinking about if we have a matrix pipeline that builds images for different arch in different Tasks - if we set this too early the Pipeline sig might still clash with other Tasks.

I think in this case, each of those Tasks produces a different image (unique digest). So the Tasks themselves wouldn't interfere with each other's signatures and attestations.

> Maybe we invert this relationship and have objects query for their parent status?

Interesting idea. It would mean that signatures and attestations may have a perceivable delay in being created (maybe?). Not sure if that's an issue.

Reply:
> In reality, we just need to wait until the last Task within the Pipeline that has signable subjects. We could enhance the logic to handle that, not sure how much more beneficial that would be though.

Makes sense! This is just waiting until 1 Task is done before annotating the PipelineRun to go though, right? That's why I'm wondering if we should just wait until the pipeline clears first before we handle the TaskRuns 🤔 

> I think in this case, each of those Tasks produces a different image (unique digest). So the Tasks themselves wouldn't interfere with each other's signatures and attestations.

Hmmm I think I'm not fully grasping the problem then - if TaskRuns are pushing unique digests that won't interfere with each other, how does this interfere with the PipelineRun? 👀 

> Interesting idea. It would mean that signatures and attestations may have a perceivable delay in being created (maybe?). Not sure if that's an issue.

I'm not super concerned - chains is already async so there's an expectation of some delay. We could use https://pkg.go.dev/knative.dev/pkg/controller#Impl.EnqueueKey to force an immediate requeue of sub resources to help speed things up if needed.

Reply:
> Makes sense! This is just waiting until 1 Task is done before annotating the PipelineRun to go though, right? 

This updates the PipelineRun whenever a TaskRun finishes. So if the PipelineRun contains 5 TaskRuns, the annotation on the PipelineRun will get updated 5 times.

> That's why I'm wondering if we should just wait until the pipeline clears first before we handle the TaskRuns

I think to make this work, we would need to update an annotation on the TaskRun to indicate whether or not it can be proceed by Chains, right? That might work. I'm not sure what the main benefit would be though. Can you help me understand?

> Hmmm I think I'm not fully grasping the problem then - if TaskRuns are pushing unique digests that won't interfere with each other, how does this interfere with the PipelineRun?

Because if you want a PipelineRun attestation for each one of those images, they need to become Pipeline results. Let's say a PipelineRun contains three TaskRuns, each building a unique image (e.g. multi-arch). Since they're unique, they have different digests, so even if they're pushed to the same OCI repo, they'll each have their own signature and attestation images. (This is also the case prior to TEP-84). Now let's say the PipelineRun produces multiple images in its result, which makes sense since at the scope of a PipelineRun, multiple images were built. The PipelineRun reconciler is going to generate the PipelineRun attestation (which is the same for all three images), and push them not only to the same OCI repo, but also to the same `sha256-<digest>.att` tag. cosign is smart enough to not overwrite the attestations that are already there, so you end up with attestation images that contain two layers. The first layer contains the TaskRun attestation (unique across all three images), and the second containing the PipelineRun attestation (the same across all three images).

Reply:
Okay, now that I'm caught up and grok everything, I think we can actually do this with less changes!

> I think to make this work, we would need to update an annotation on the TaskRun to indicate whether or not it can be proceed by Chains, right? That might work. I'm not sure what the main benefit would be though. Can you help me understand?

I think we already have this in the `chains.tekton.dev/signed` annotation + the reconciled check here: https://github.com/tektoncd/chains/blob/ad5a650eca7cc0d3019d5712bc945f906c2a4d1b/pkg/chains/annotations.go#L36-L43

I think what we probably want to do is swap out the direct client in https://github.com/tektoncd/chains/blob/ad5a650eca7cc0d3019d5712bc945f906c2a4d1b/pkg/reconciler/pipelinerun/pipelinerun.go#L94 for the lister cache via the informer from https://pkg.go.dev/github.com/tektoncd/pipeline@v0.40.0/pkg/client/injection/informers/pipeline/v1beta1/taskrun . This should be safe to query as much as we want and should generally have the latest-ish values.

To trigger the resync from the TaskRun, you can just requeue the parent object key using https://pkg.go.dev/knative.dev/pkg@v0.0.0-20220921024409-d1d5c849073b/controller#Impl.EnqueueKey

This should block the PipelineRun reconcile until all the TaskRuns are done and minimize the # of API calls the PipelineRun reconciler is making to query TaskRun state.

> The reason I chose the current approach is because I wanted to reconcile resources as early as possible. Waiting to reconcile TaskRuns until the whole PipelineRun completes goes against that. I'm not saying that we shouldn't do that, just clarifying my thought process.

Got it. Makes sense!

Reply:
ooh I like this idea.

I replaced the line with:
```go
taskruninformer.Get(ctx).Lister().TaskRuns(pr.Namespace).Get(name)
```

But I'm getting a panic when that line is run:
```go
Unable to fetch github.com/tektoncd/pipeline/pkg/client/informers/externalversions/pipeline/v1beta1.TaskRunInformer from context.
```

I'm not well-versed here, but it sounds like maybe I also need to do some setup on the informer.

Reply:
Yeah I always forget where the client injection needs to go. Try passing it when the controller is created similar to the pipelinerun informer. https://github.com/tektoncd/chains/blob/ad5a650eca7cc0d3019d5712bc945f906c2a4d1b/pkg/reconciler/pipelinerun/controller.go#L33

Reply:
Thanks for the tip :pray:  Pushed a new commit that takes this approach. It should further improve performance.

For the second part of your suggestion, I'm not sure if it'll work. It would require the TaskRun controller/reconciler to add something on the queue of the PipelineRun controller/reconciler. I'm likely missing something obvious here though.

Reply:
Shout out to @vaikas for this!

We should be able to use a tracker for this to watch a PipelineRun's children:

https://github.com/sigstore/policy-controller/blob/main/pkg/reconciler/clusterimagepolicy/clusterimagepolicy.go#L204 for an example

Reply:
Kudos to @vaikas!

Added a commit that takes this approach.

@wlynch, if the current changes look good to you, I can squash the commits if you prefer.

(The failing e2e test seems to be unrelated to these changes because it's also failing on my PR that updates just the docs.)